File: lib/rails/generators/channel/templates/channel.js
import consumer from "./consumer"

// Channel factory function - call this to create a connection when needed
function create<%= channel_name.gsub('Channel', '') %>Channel(params = {}) {
  // Store params for use in callbacks
  const subscriptionParams = { ...params };
  
  return consumer.subscriptions.create(
    {
      channel: "<%= channel_name %>",
      ...params<% if actions.include?('subscribed') %>,
      // Add any parameters needed for subscription
      // room_id: document.querySelector('[data-room-id]')?.dataset.roomId<% end %>
    },
    {
    connected() {
      // Called when the subscription is ready for use on the server
      console.log("Connected to <%= channel_name %>");
    },

    disconnected() {
      // Called when the subscription has been terminated by the server
      console.log("Disconnected from <%= channel_name %>");
      
      // 创建详细的断开连接错误信息
      const disconnectionError = {
        type: 'error',
        message: '<%= channel_name %> connection was disconnected',
        channel: '<%= channel_name %>',
        params: subscriptionParams,
        details: {
          connectionType: 'ActionCable WebSocket',
          event: 'disconnected',
          suggestion: 'Connection lost - check network or server status or connection identity'
        }
      };
      this.handleGlobalError(disconnectionError);
    },

    received(data) {
      // Called when there's incoming data on the websocket for this channel
      console.log("Received data:", data);

      // Restore any disabled buttons when receiving ActionCable messages
      if (typeof window.restoreButtonStates === 'function') {
        window.restoreButtonStates();
      }

      // Global error handling
      if (data.type === 'error' && data.success === false) {
        this.handleGlobalError(data);
        return;
      }

      // Handle different types of messages
      switch(data.type) {
<% actions.each do |action| %>
        case '<%= action %>':
          this.handle<%= action.camelize %>(data);
          break;
<% end %>
        default:
          console.log("Unknown message type:", data.type);
      }
    },

    // Global error handling for this channel
    handleGlobalError(errorData) {
      // Send to global error handler
      if (window.errorHandler && window.errorHandler.handleActionCableError) {
        window.errorHandler.handleActionCableError({
          ...errorData,
        });
      } else {
        // Fallback error handling
        console.error('Channel Error:', {
          ...errorData
        });
      }
    },

<% actions.each do |action| %>
    // Send <%= action %> to the server
    <%= action.camelize(:lower) %>(data) {
      this.perform('<%= action %>', data);
    },

    // Handle <%= action %> message from server
    handle<%= action.camelize %>(data) {
      // Implement your <%= action %> handling logic here
      console.log('<%= action.camelize %> received:', data);
    },

<% end %>
    }
  );
}

// Make factory function available globally for easy access in views
window.create<%= channel_name.gsub('Channel', '') %>Channel = create<%= channel_name.gsub('Channel', '') %>Channel;

// Export the factory function
export default create<%= channel_name.gsub('Channel', '') %>Channel;

// Optional: Create a default instance (uncomment if you want auto-connection)
// const <%= javascript_channel_name %>Channel = create<%= channel_name.gsub('Channel', '') %>Channel();
// window.<%= javascript_channel_name %>Channel = <%= javascript_channel_name %>Channel;
// export default <%= javascript_channel_name %>Channel;

File: lib/rails/generators/channel/templates/channel.rb
class <%= channel_name %> < ApplicationCable::Channel
  def subscribed
    # Stream from a channel based on some identifier
    # Example: stream_from "some_channel"
    stream_from "<%= stream_name %>"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

<% actions.each do |action| %>
  def <%= action %>(data)
    # Handle <%= action %> action
    # You can access:
    # - data: the data sent from client
    # - current_user: the authenticated user (if available)
    # - params: parameters passed when subscribing
    
    # Example broadcast:
    # ActionCable.server.broadcast(
    #   "<%= stream_name %>",
    #   {
    #     type: '<%= action %>',
    #     data: data,
    #     user: current_user&.slice(:id, :name, :email)
    #   }
    # )
  end

<% end %>
  private

<% if requires_authentication? -%>
  def current_user
    @current_user ||= connection.current_user
  end
<% else -%>
  # def current_user
  #   @current_user ||= connection.current_user
  # end
<% end -%>
end
